students/Ben_freeman/lesson08/Circle_test.py

```diff
@@ -0,0 +1,112 @@
+from Circle_class_exercise import Circle, Sphere
+import random
+
+
+def test_radius():
+    c = Circle(4)
+    print(c.radius)
+    assert c.radius == 4
+
+
+def test_diameter():
+    c = Circle(4)
+    print(c.diameter)
+    assert c.diameter == 8
+
+
+def test_setter_diameter():
+    c = Circle(4)
+    c.diameter = 2
+    print(c.diameter, "and", c.radius)
+    assert c.radius == 1
+
+
+def test_area():
+    c = Circle(2)
+    assert round(c.area,5) == 12.566370
+
+
+def test_area_Setter():
+    c = Circle(2)
+    try:
+        c.area = 42
+    except AttributeError:
+        assert True
+
+
+def test_from_diameter():
+    c = Circle.from_diameter(8)
+    print(c.radius)
+    assert c.radius == 4
+    assert c.diameter == 8
+
+
+def test_str():
+    c = Circle(4)
+    assert print(Circle(4)) == "Circle with radius: 4"
```

Comment:
Is there a way to control what type of object is returned?

Reply:
I don't fully understand the question - you mean control what print() returns?

Reply:
I'm mostly just kinda confused about what is actually happening here I guess.

`In [1]: run Circle_class_exercise.py

In [2]: c = Circle(4)

In [3]: print(c)
Circle with radius: 4

In [4]: type(print(c))
Circle with radius: 4
Out[4]: NoneType`

I think what is confusing me is that it seems like when I input print(c), it seems like I get a string back, but its actually NoneType when I check it out. In my original question I was mostly just wondering if I could change something so that 
`type(print(c))`
would return
`str`

Now I'm mostly just curious what is actually going on when python prints "Circle with radius:4" since its not a string thats getting returned to me.

Reply:
I see. What you actually want for your unit test is `assert str(Circle(4)) == "Circle with radius: 4"`

your `__str__` magic function is to support str() method.


The relation of print() to `__str__` is that it will first convert your object to string via str() which is why you are implementing that method, besides it has nothing to do with str() except using it.

the print() itself is writing "a string" to standard out via sys.stdout.write